File: tui/src/scala/tui/widgets/GaugeWidget.scala
package tui
package widgets

import tui.Style
import tui.internal.ranges

/** A widget to display a task progress.
  */
case class GaugeWidget(
    block: Option[BlockWidget] = None,
    ratio: GaugeWidget.Ratio = GaugeWidget.Ratio.Zero,
    label: Option[Span] = None,
    useUnicode: Boolean = false,
    style: Style = Style.DEFAULT,
    gaugeStyle: Style = Style.DEFAULT
) extends Widget {

  override def render(area: Rect, buf: Buffer): Unit = {
    buf.setStyle(area, style)
    val gauge_area = block match {
      case Some(b) =>
        val inner_area = b.inner(area)
        b.render(area, buf)
        inner_area

      case None => area
    }
    buf.setStyle(gauge_area, gaugeStyle)
    if (gauge_area.height < 1) {
      return
    }

    // compute label value and its position
    // label is put at the center of the gauge_area
    val label = {
      val pct = math.round(ratio.value * 100.0)
      this.label.getOrElse(Span.nostyle(pct.toString + "%"))
    }
    val clamped_label_width = gauge_area.width.min(label.width)
    val label_col = gauge_area.left + (gauge_area.width - clamped_label_width) / 2
    val label_row = gauge_area.top + gauge_area.height / 2

    // the gauge will be filled proportionally to the ratio
    val filled_width = gauge_area.width.toDouble * this.ratio.value
    val end: Int = if (useUnicode) {
      gauge_area.left + math.floor(filled_width).toInt
    } else {
      gauge_area.left + math.round(filled_width).toInt
    }
    ranges.range(gauge_area.top, gauge_area.bottom) { y =>
      // render the filled area (left to end)
      ranges.range(gauge_area.left, end) { x =>
        // spaces are needed to apply the background styling
        buf
          .get(x, y)
          .setSymbol(" ")
          .setFg(gaugeStyle.bg.getOrElse(Color.Reset))
          .setBg(gaugeStyle.fg.getOrElse(Color.Reset))
        ()
      }
      if (useUnicode && ratio.value < 1.0) {
        buf
          .get(end, y)
          .setSymbol(getUnicodeBlock(filled_width % 1.0))
        ()
      }
    }
    // set the span
    buf.setSpan(label_col, label_row, label, clamped_label_width)
    ()
  }

  def getUnicodeBlock(frac: Double): String =
    math.round(frac * 8.0) match {
      case 1 => symbols.block.ONE_EIGHTH
      case 2 => symbols.block.ONE_QUARTER
      case 3 => symbols.block.THREE_EIGHTHS
      case 4 => symbols.block.HALF
      case 5 => symbols.block.FIVE_EIGHTHS
      case 6 => symbols.block.THREE_QUARTERS
      case 7 => symbols.block.SEVEN_EIGHTHS
      case 8 => symbols.block.FULL
      case _ => " "
    }
}

object GaugeWidget {
  case class Ratio(value: Double) {
    require(value >= 0 && value <= 1, s"$value is not between 0 and 1")
  }

  object Ratio {
    val Zero = new Ratio(0.0)

    def percent(value: Double): Ratio =
      new Ratio(value / 100)
  }
}

File: tui/src/java/tui/widgets/GaugeWidget.java
package tui.widgets;

import java.util.Optional;
import tui.Buffer;
import tui.Color;
import tui.Rect;
import tui.Span;
import tui.Style;
import tui.Symbols;
import tui.Widget;
import tui.internal.Ranges;

/// A widget to display a task progress.
public final class GaugeWidget implements Widget {
  public final Optional<BlockWidget> block;
  public final Ratio ratio;
  public final Optional<Span> label;
  public final boolean useUnicode;
  public final Style style;
  public final Style gaugeStyle;

  public GaugeWidget(
      Optional<BlockWidget> block,
      Ratio ratio,
      Optional<Span> label,
      boolean useUnicode,
      Style style,
      Style gaugeStyle) {
    this.block = block;
    this.ratio = ratio;
    this.label = label;
    this.useUnicode = useUnicode;
    this.style = style;
    this.gaugeStyle = gaugeStyle;
  }

  public static GaugeWidget empty() {
    return new GaugeWidget(
        Optional.empty(), Ratio.Zero, Optional.empty(), false, Style.DEFAULT, Style.DEFAULT);
  }

  public GaugeWidget withBlock(BlockWidget b) {
    return new GaugeWidget(Optional.of(b), ratio, label, useUnicode, style, gaugeStyle);
  }

  public GaugeWidget withRatio(Ratio r) {
    return new GaugeWidget(block, r, label, useUnicode, style, gaugeStyle);
  }

  public GaugeWidget withLabel(Span l) {
    return new GaugeWidget(block, ratio, Optional.of(l), useUnicode, style, gaugeStyle);
  }

  public GaugeWidget withUseUnicode(boolean u) {
    return new GaugeWidget(block, ratio, label, u, style, gaugeStyle);
  }

  public GaugeWidget withStyle(Style s) {
    return new GaugeWidget(block, ratio, label, useUnicode, s, gaugeStyle);
  }

  public GaugeWidget withGaugeStyle(Style s) {
    return new GaugeWidget(block, ratio, label, useUnicode, style, s);
  }

  @Override
  public void render(Rect area, Buffer buf) {
    buf.setStyle(area, style);
    Rect gaugeArea;
    if (block.isPresent()) {
      BlockWidget b = block.get();
      Rect innerArea = b.inner(area);
      b.render(area, buf);
      gaugeArea = innerArea;
    } else {
      gaugeArea = area;
    }
    buf.setStyle(gaugeArea, gaugeStyle);
    if (gaugeArea.height() < 1) {
      return;
    }

    long pct = Math.round(ratio.value * 100.0);
    Span labelVal = label.orElseGet(() -> Span.nostyle(pct + "%"));
    int clampedLabelWidth = Math.min(gaugeArea.width(), labelVal.width());
    int labelCol = gaugeArea.left() + (gaugeArea.width() - clampedLabelWidth) / 2;
    int labelRow = gaugeArea.top() + gaugeArea.height() / 2;

    double filledWidth = (double) gaugeArea.width() * ratio.value;
    int end;
    if (useUnicode) {
      end = gaugeArea.left() + (int) Math.floor(filledWidth);
    } else {
      end = gaugeArea.left() + (int) Math.round(filledWidth);
    }
    Ranges.range(
        gaugeArea.top(),
        gaugeArea.bottom(),
        y -> {
          Ranges.range(
              gaugeArea.left(),
              end,
              x -> {
                // Use full block for the filled part of the gauge and spaces for the part
                // that's covered by the label. The bg/fg are swapped for the label part so
                // the label remains legible against the filled background.
                if (x < labelCol || x > labelCol + clampedLabelWidth || y != labelRow) {
                  buf.get(x, y)
                      .setSymbol(Symbols.block.FULL)
                      .setFg(gaugeStyle.fg().orElse(Color.Reset))
                      .setBg(gaugeStyle.bg().orElse(Color.Reset));
                } else {
                  buf.get(x, y)
                      .setSymbol(" ")
                      .setFg(gaugeStyle.bg().orElse(Color.Reset))
                      .setBg(gaugeStyle.fg().orElse(Color.Reset));
                }
              });
          if (useUnicode && ratio.value < 1.0) {
            buf.get(end, y).setSymbol(getUnicodeBlock(filledWidth % 1.0));
          }
        });

    buf.setSpan(labelCol, labelRow, labelVal, clampedLabelWidth);
  }

  public static String getUnicodeBlock(double frac) {
    long v = Math.round(frac * 8.0);
    if (v == 1) return Symbols.block.ONE_EIGHTH;
    if (v == 2) return Symbols.block.ONE_QUARTER;
    if (v == 3) return Symbols.block.THREE_EIGHTHS;
    if (v == 4) return Symbols.block.HALF;
    if (v == 5) return Symbols.block.FIVE_EIGHTHS;
    if (v == 6) return Symbols.block.THREE_QUARTERS;
    if (v == 7) return Symbols.block.SEVEN_EIGHTHS;
    if (v == 8) return Symbols.block.FULL;
    return " ";
  }

  public record Ratio(double value) {
    public Ratio {
      if (value < 0 || value > 1) {
        throw new IllegalArgumentException(value + " is not between 0 and 1");
      }
    }

    public static final Ratio Zero = new Ratio(0.0);

    public static Ratio percent(double value) {
      return new Ratio(value / 100);
    }
  }
}
